reconciler/namespace: implement location API MVP

sttts opened the pull request:
Depends on https://github.com/kcp-dev/kcp/pull/1082 and https://github.com/kcp-dev/kcp/pull/1104.

- implement (MVP-)complete location-based placement including extensive unit tests
- deprecated `--auto-publish-apis` and always set to false
- with the previous change, switch local TMC to APIExports and APIBindings
- to make local TMC easy, auto-create `"kubernetes"` APIExport and local APIBinding when either Locations or WorkloadCluster are found (controller workload/apiexportcreate)
- restrict workloads/apiexport controller to `"kubernetes"` APIExport (fixes https://github.com/kcp-dev/kcp/issues/1088)
- adapt workload namespace and resource controller
- drop legacy scheduling (via `workloads.kcp.dev/cluster`).
- disabled ingresssplitter test. The splitter controller is conceptually not able to watch APIBinding resource like services and ingresses. It has to watch per compute service.
- fix all other e2e tests
- enable KCPLocationAPI feature gate by default

Follow-ups:
- [ ] v0.5: allow cross-org APIBindings https://github.com/kcp-dev/kcp/pull/1125
- [ ] v0.6: switch placement annotations and workload scheduling state to UIDs
- [ ] v0.6: fix/rewrite ingress-splitter
- [ ] v0.6: project locations into binding workspaces

itdove commented on pkg/reconciler/workload/resource/resource_controller.go:
```diff
@@ -0,0 +1,355 @@
+			UpdateFunc: func(old, obj interface{}) {
+				oldNS := old.(*corev1.Namespace)
+				newNS := obj.(*corev1.Namespace)
+				if !reflect.DeepEqual(scheduleStateLabels(oldNS.Labels), scheduleStateLabels(newNS.Labels)) {
```
ns.Annotations should be used instead of labels as the placement_controller use annotations and not label.
The use of ns,label is accross the board and so needs to be changed in several places.

sttts replied:
if I am not mistaken, these are different keys: `schedulingv1alpha1.PlacementAnnotationKey` is an annotation, `workloadv1alpha1.InternalClusterResourceStateLabelPrefix` and `namespace.DeprecatedScheduledClusterNamespaceLabel` are about labels.

sttts replied:
fixed

sttts commented:
This needs https://github.com/kcp-dev/kcp/pull/1082 because the resource controller otherwise cannot see APIBinding objects. Hence, the syncer will never see those either.

ncdc commented:
I think #1082 is working. I've added/updated e2es for it. Would be interested to know what you think.

sttts commented:
> I think https://github.com/kcp-dev/kcp/pull/1082 is working. I've added/updated e2es for it. Would be interested to know what you think.

Rebased onto it. And it works 🎉

stevekuznetsov commented on pkg/reconciler/workload/resource/resource_reconcile_test.go:
```diff
@@ -0,0 +1,187 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gotAnnotationPatch, gotLabelPatch := computePlacement(tt.ns, tt.obj)
+			if !equality.Semantic.DeepEqual(gotAnnotationPatch, tt.wantAnnotationPatch) {
```
Can we please always use the following on larger objects

```go
if diff := cmp.Diff(gotAnnotationPatch, tt.wantAnnotationPatch); diff != "" {
    t.Errorf("incorrect annotation patch: %s", diff)
}
```

sttts replied:
❤️

stevekuznetsov reviewed (commented):
Made it part way through ... this is entirely too much to review. I suggest we split this up into separate bits like:

 - add reconcilers for namespaces & objects to have the labels we expect, but don't implement rescheduling
 - add rescheduling
 - add apiexport for kubernetes 
 
 etc

There's so much intricate logic in each one of these bits that reviewing all together is just impossible - so many branches/edges ...

sttts commented on pkg/reconciler/workload/namespace/namespace_reconcile.go:
```diff
@@ -97,19 +103,160 @@ func (c *Controller) reconcileNamespace(ctx context.Context, lclusterName logica
+// ensureNamespaceScheduled will ensure every placement to a location is properly represented in the
+// state.workloads.kcp.dev/* labels.
+func (c *Controller) ensureNamespaceScheduled(ctx context.Context, ns *corev1.Namespace, placement schedulingv1alpha1.PlacementAnnotation) (*corev1.Namespace, bool, error) {
+	expectedPlacement := map[string]interface{}{}
```
nil = delete

sttts replied:
doc'ed

sttts commented on pkg/reconciler/workload/namespace/namespace_reconcile.go:
```diff
@@ -97,19 +103,160 @@ func (c *Controller) reconcileNamespace(ctx context.Context, lclusterName logica
+			return ns, false, nil
+		}
+	}
+	patchBytes, err := json.Marshal(patch)
```
nil = delete, value = upsert

sttts commented on pkg/reconciler/workload/resource/resource_reconcile.go:
```diff
@@ -113,24 +160,12 @@ func (c *Controller) reconcileGVR(gvr schema.GroupVersionResource) error {
-		return "", nil, err
-	}
-	return types.MergePatchType, bs, nil
+func toJSON(value interface{}) string {
```
used in logs only

sttts replied:
fixed

sttts commented on pkg/reconciler/scheduling/location/workloadcluster_helpers.go:
```diff
@@ -47,11 +48,23 @@ func LocationWorkloadClusters(workloadClusters []*workloadv1alpha1.WorkloadClust
+
+// FilterNonEvicting filters out the evicing workload clusters.
+func FilterNonEvicting(workloadClusters []*workloadv1alpha1.WorkloadCluster) []*workloadv1alpha1.WorkloadCluster {
+	ret := make([]*workloadv1alpha1.WorkloadCluster, 0, len(workloadClusters))
```
Evicting-vs-non-Evicting is like 1-to-99. I think 1% over-allocation is fine.

sttts commented on pkg/reconciler/workload/resource/resource_reconcile_test.go:
```diff
@@ -0,0 +1,187 @@
+			obj: object(map[string]string{
+				"deletion.internal.workloads.kcp.dev/cluster-3": "2002-10-02T10:00:00-05:00",
+			}, map[string]string{
+				"state.internal.workloads.kcp.dev/cluster-3": "Sync", // removed hard
```
clarified

sttts commented on pkg/reconciler/workload/resource/resource_reconcile_test.go:
```diff
@@ -0,0 +1,187 @@
+				"state.internal.workloads.kcp.dev/cluster-1": "Sync",
+			},
+		},
+		{name: "new location",
```
it was not. Fixed.

sttts commented on pkg/reconciler/workload/resource/resource_reconcile_test.go:
```diff
@@ -0,0 +1,187 @@
+		wantAnnotationPatch map[string]interface{}
+		wantLabelPatch      map[string]interface{}
+	}{
+		{name: "nothing",
```
improved the strings

sttts commented on pkg/reconciler/workload/resource/resource_reconcile.go:
```diff
@@ -33,76 +35,121 @@ import (
+	}
+
+	// create merge patch
+	annotationPatch = map[string]interface{}{}
```
Good point. Made it a string pointers, Golang's poor men's `Option[string]`.

sttts commented:
@stevekuznetsov addressed your comments

sttts commented:
> I suggest we split this up into separate bits like:
> add reconcilers for namespaces & objects to have the labels we expect, but don't implement rescheduling
> add rescheduling
> add apiexport for kubernetes
> There's so much intricate logic in each one of these bits that reviewing all together is just impossible - so many branches/edges ...

We talked about it in Slack. This PR here replaces basically all of scheduling. Having it partially merge would mean a lot more complexity in the intermediate steps. E.g. we cannot switch to APIExports without having them for local scheduling. Rescheduling is needed for existing eviction e2e tests. Placement, namespace, resource scheduling only make sense together. Having a second scheduling in the namespace controller is pain, I went that route first, the commits before squashing showed that path, not worth it.

Also note that the PR is now much smaller with the pre-reqs merged. Finally note that lots of the LoCs is due to new test coverage (we had zero unit tests in large parts of the scheduling code).

shawn-hurley reviewed (commented):
Just trying to help with reviews

sttts commented on pkg/reconciler/scheduling/placement/placement_controller.go:
```diff
@@ -157,8 +166,44 @@ func NewController(
+
+				// compare ignoring heart-beat
+				if !reflect.DeepEqual(oldClusterCopy, newClusterCopy) {
+					c.enqueueWorkloadCluster(obj)
```
look into the enqueue functions. All eventually put namespaces into the queue only.

Rule of thumb: only one resource type in the queue. Here that are namespaces.

sttts commented on pkg/reconciler/scheduling/placement/placement_reconcile_scheduling.go:
```diff
@@ -0,0 +1,426 @@
+	if len(workloadBindings) == 0 {
+		// TODO(sttts): we have no good way to identify workload APIBinding right now. So we use some quite high
+		//              requeue duration. But better than not requeueing at all.
+		r.enqueueAfter(clusterName, ns, time.Minute*2)
```
we should. updated

sttts commented on pkg/reconciler/scheduling/placement/placement_reconcile_scheduling.go:
```diff
@@ -0,0 +1,426 @@
+		switch state {
+		case schedulingv1alpha1.PlacementStatePending, schedulingv1alpha1.PlacementStateBound:
+			activeLocations.Insert(locationName)
+			locationsToSchedule.Delete(locationName)
```
yes. Renamed.

sttts commented on pkg/reconciler/scheduling/placement/placement_reconcile_scheduling.go:
```diff
@@ -0,0 +1,426 @@
+	var lastErr error
+	admissiblePlacements := sets.NewString()
+	readyPlacements := sets.NewString()
+	readyLocationClusters := map[string]sets.String{}
```
yes. Had moved that readiness logic into the place function. Fixed
